src/SIL.Harmony/SnapshotWorker.cs

```diff
@@ -223,6 +196,35 @@ internal async IAsyncEnumerable<ObjectSnapshot> GetSnapshotsWhere(Expression<Fun
         }
     }
 
+    private async Task GenerateSnapshotForEntity(IObjectBase entity, ObjectSnapshot? prevSnapshot, ChangeContext context)
+    {
+        //to get the state in a point in time we would have to find a snapshot before that time, then apply any commits that came after that snapshot but still before the point in time.
+        //we would probably want the most recent snapshot to always follow current, so we might need to track the number of changes a given snapshot represents so we can
+        //decide when to create a new snapshot instead of replacing one inline. This would be done by using the current snapshots parent, instead of the snapshot itself.
+        // s0 -> s1 -> sCurrent
+        // if always taking snapshots would become
+        // s0 -> s1 -> sCurrent -> sNew
+        //but but to not snapshot every change we could do this instead
+        // s0 -> s1 -> sNew
+
+        //when both snapshots are for the same commit we don't want to keep the previous, therefore the new snapshot should be root
+        var isRoot = prevSnapshot is null || (prevSnapshot.IsRoot && prevSnapshot.CommitId == context.Commit.Id);
+        var newSnapshot = new ObjectSnapshot(entity, context.Commit, isRoot);
+        //if both snapshots are for the same commit then we don't want to keep the previous snapshot
+        if (prevSnapshot is null || prevSnapshot.CommitId == context.Commit.Id)
+        {
+            //do nothing, will cause prevSnapshot to be overriden in _pendingSnapshots if it exists
+        }
+        else if (context.CommitIndex % 2 == 0 && !prevSnapshot.IsRoot && IsNew(prevSnapshot))
+        {
+            context.IntermediateSnapshots[prevSnapshot.Entity.Id] = prevSnapshot;
```

Comment:
I'll have to think about this as I don't really like it due to how indirect it is, it was fine before as the variable never left the scope of that function. Really I never liked having that intermediate snapshots variable in the first place, and this just pushed it to more uncomfortable...

Reply:
Yeah, I agree.
It would probably make sense to defer deciding which snapshots to actualy persist until after `ApplyCommitChanges`.

Reply:
I started work on this https://github.com/sillsdev/harmony/pull/34
Would you be ok merging this as it is and refactoring in the follow up or do you want me to revert adding intermediate-Snapshots to the change context?